pkg/sql/sqlbase/structured.go

```diff
@@ -1886,6 +1886,29 @@ func (desc *TableDescriptor) FindActiveColumnByID(id ColumnID) (*ColumnDescripto
 	return nil, fmt.Errorf("column-id \"%d\" does not exist", id)
 }
 
+// FindReadableColumnByID finds the column with specified ID and also
+// sets it to nullable if it's in the middle of a schema change.
+func (desc *TableDescriptor) FindReadableColumnByID(id ColumnID) (ColumnDescriptor, error) {
+	for i, c := range desc.Columns {
+		if c.ID == id {
+			return desc.Columns[i], nil
+		}
+	}
+	for _, m := range desc.Mutations {
+		if c := m.GetColumn(); c != nil {
+			if c.ID == id {
+				col := *c
+				// Even if the column is non-nullable it can be null in the
+				// middle of a schema change while it's being baxckfilled
+				// to a non-nullable value.
+				col.Nullable = true
```

Comment:
@vivekmenezes I think we need to extend this change. Let me explain.

The change you are making here appears to help the row **reader** to not be confused when it sees a null value in a non-nullable column. I have some comments about this (see point 2 below) but let's keep it at that for now.

1) The main issue here is that changing this property will create a semantic problem for **row writers** -- it is important that a client app **receives a SQL error if they attempt to insert a NULL value into a non-NULLable column.** With your change here, I believe this check is not done any more.

So I think the first order of business is to verify (with tests) that NULL inserts into non-NULLable columns currently undergoing a backfill are still/properly rejected by SQL errors. 

2) regarding the point at the top, with the row readers. In any case it is invalid for a SQL client to receive a NULL value as response for a query to a non-NULL column. We must not allow this to happen in any case -- this is a plain correctness error, a silent data error (S-0). 

I wonder if your change here makes it more or less likely to encounter this situation. If the column is marked as nullable in the descriptor, and there is a query to read a row in the process of being added, the correct behavior is this:

   2.a) the query must not see the column at all -- it must operate on the table as if the column did not exist
   2.b) in any case, there must not be any NULL value visible by the query

In contrast, the following behavior would be incorrect:

   2.c) the query sees the column
   2.d) the query sees a NULL value in the column


What would you say in response to my comments here?

Reply:
Also to ensure the new added column is not visible it will be interesting to ensure:

1) we have a test that the following statement groups each produce an error:

```sql
CREATE TABLE t (x INT PRIMARY KEY);
BEGIN;
ALTER TABLE t ADD COLUMN y INT NOT NULL DEFAULT 123;
INSERT INTO t (x) VALUES (1) ON CONFLICT DO UPDATE SET x = t.y + 1; -- this must error with "t.y is not public"
COMMIT;
```

2) that the behavior with stored columns that depend on both new columns and existing columns is clear:

```sql
CREATE TABLE t (x INT PRIMARY KEY);
BEGIN;
ALTER TABLE t ADD COLUMN y INT NOT NULL DEFAULT 123,
              ADD COLUMN z INT AS (x + y) STORED;
INSERT INTO t (x) VALUES (2); --  see note below
COMMIT;
```

In this example the column `z` must be updated both by the backfill _and_ the insert, so that the final value of `z` is properly equal to 125.